lib/Runtime/Library/JavascriptSet.cpp

```diff
@@ -6,379 +6,669 @@
-        {
-            JavascriptError::ThrowTypeErrorVar(scriptContext, JSERR_NeedObjectOfType, _u("Set.prototype.entries"), _u("Set"));
-        }
+    if (!this->u.simpleVarSet->ContainsKey(simpleVar))
+    {
+        SetDataNode* node = this->list.Append(simpleVar, GetScriptContext()->GetRecycler());
+        this->u.simpleVarSet->Add(simpleVar, node);
+    }
+
+    return true;
+}
 
-        JavascriptSet* set = JavascriptSet::FromVar(args[0]);
+void
+JavascriptSet::AddToComplexVarSet(Var value)
+{
+    if (!this->u.complexVarSet->ContainsKey(value))
+    {
+        SetDataNode* node = this->list.Append(value, GetScriptContext()->GetRecycler());
+        this->u.complexVarSet->Add(value, node);
+    }
+}
 
-        return scriptContext->GetLibrary()->CreateSetIterator(set, JavascriptSetIteratorKind::KeyAndValue);
+void
+JavascriptSet::Add(Var value)
+{
+    switch (this->kind)
+    {
+    case SetKind::EmptySet:
+    {
+        this->AddToEmptySet(value);
+        return;
     }
 
-    Var JavascriptSet::EntryValues(RecyclableObject* function, CallInfo callInfo, ...)
+    case SetKind::IntSet:
     {
-        PROBE_STACK(function->GetScriptContext(), Js::Constants::MinStackDefault);
+        if (this->TryAddToIntSet(value))
+        {
+            return;
+        }
+
+        Var simpleVar = JavascriptConversion::TryCanonicalizeAsSimpleVar(value);
+        if (simpleVar)
+        {
+            this->PromoteToSimpleVarSet();
+            this->Add(simpleVar);
+            return;
+        }
 
-        ARGUMENTS(args, callInfo);
-        ScriptContext* scriptContext = function->GetScriptContext();
+        this->PromoteToComplexVarSet();
+        this->Add(value);
+        return;
+    }
 
-        if (!JavascriptSet::Is(args[0]))
+    case SetKind::SimpleVarSet:
+        if (this->TryAddToSimpleVarSet(value))
         {
-            JavascriptError::ThrowTypeErrorVar(scriptContext, JSERR_NeedObjectOfType, _u("Set.prototype.values"), _u("Set"));
+            return;
         }
 
-        JavascriptSet* set = JavascriptSet::FromVar(args[0]);
+        this->PromoteToComplexVarSet();
+        this->Add(value);
+        return;
 
-        return scriptContext->GetLibrary()->CreateSetIterator(set, JavascriptSetIteratorKind::Value);
+    case SetKind::ComplexVarSet:
+        this->AddToComplexVarSet(value);
+        return;
+
+    default:
+        Assume(UNREACHED);
     }
+}
 
-    Var JavascriptSet::EntryGetterSymbolSpecies(RecyclableObject* function, CallInfo callInfo, ...)
+void JavascriptSet::Clear()
+{
+    // TODO: (Consider) Should we clear the set here and leave it as large as it has grown, or
+    // toss it away and create a new empty set, letting it grow as needed?
+    this->list.Clear();
+    switch (this->kind)
     {
-        ARGUMENTS(args, callInfo);
-
-        Assert(args.Info.Count > 0);
+    case SetKind::EmptySet:
+        return;
+    case SetKind::IntSet:
+        this->u.intSet->ClearAll();
```

Comment:
Did we consider changing the kind to EmptySet ?

Reply:
I considered it, but figured that often after clearing a set you will want to put things in it again, so it's cheaper to avoid extra allocation

---
In reply to: [174316471](https://github.com/Microsoft/ChakraCore/pull/4816#discussion_r174316471) [](ancestors = 174316471)

Reply:
sounds good. 

---
In reply to: [174319064](https://github.com/Microsoft/ChakraCore/pull/4816#discussion_r174319064) [](ancestors = 174319064,174316471)